### PR TITLE
Add support for reverse consumer information

### PR DESCRIPTION
**Pull request by mook-as**
In `/var/vcap/job-src/*/config_spec.json`, we will now include a list of consumers of this job's links (via the `consumed_by` key, similar to the existing `consumes` key).  This is useful so we can attempt to restart pods which contain outdated information (in case they linked against a previous iteration of this pod and now contains outdated information).

/cc @viovanov

### CODE REVIEW

File: model/job_reference.go [hunk 1]
```diff
@@ -11,7 +11,8 @@ type JobReference struct {
 	Name                string                     `yaml:"name"`    // The name of the job
 	ReleaseName         string                     `yaml:"release"` // The release the job comes from
 	ExportedProviders   map[string]JobProvidesInfo `yaml:"provides"`
-	ResolvedConsumers   map[string]JobConsumesInfo `yaml:"consumes"`
+	ResolvedConsumers   map[string]JobConsumesInfo `yaml:"consumes"`    // Instance groups that this job links to & requires
```
Comment (f0rmiga) [inline model/job_reference.go:14]: How about aligning the struct field label with the yaml field label? It is a bit confusing one being `...Consumers` and the other `consumes`.

File: model/resolver/resolver.go [hunk 2]
```diff
@@ -345,6 +345,43 @@ func (r *Resolver) ResolveLinks() validation.ErrorList {
 		}
 	}
 
+	// Record in each job what roles consume it
```
Comment (f0rmiga) [inline model/resolver/resolver.go]: [nit] Punctuation here and below.

File: model/resolver/resolver.go [hunk 3]
```diff
@@ -345,6 +345,43 @@ func (r *Resolver) ResolveLinks() validation.ErrorList {
...
+	for _, consumerInstanceGroup := range m.InstanceGroups {
```
Comment (f0rmiga) [inline model/resolver/resolver.go:372]: What do you think about pulling the logic from the inner-most loop into a function?

File: model/resolver/resolver.go [hunk 4]
```diff
@@ -345,6 +345,43 @@ func (r *Resolver) ResolveLinks() validation.ErrorList {
...
+		for _, consumerJob := range consumerInstanceGroup.JobReferences {
+			for linkName, consumer := range consumerJob.ResolvedConsumers {
+				providerInstanceGroup := m.LookupInstanceGroup(consumer.RoleName)
+				if providerInstanceGroup == nil {
+					// This should not happen: we resolved a link, but can no
+					// longer find the instance group that provides it
+					errors = append(errors, validation.InternalError(
+						fmt.Sprintf(`instance_group[%s].job[%s].consumes[%s]`, consumerInstanceGroup.Name, consumerJob.Name, linkName),
```
Comment (f0rmiga) [inline model/resolver/resolver.go]: Declaring variables in a more procedural approach would help reading it. E.g.:
```go
field := fmt.Sprintf("instance_group[%s].job[%s].consumes[%s]", consumerInstanceGroup.Name, consumerJob.Name, linkName)
err := fmt.Errorf("Could not find resolved instance group %s", consumer.RoleName)
internalErr := validation.InternalError(field, err)
errors = append(errors, internalErr)
```

Also, no need for using ` for the strings.

File: model/resolver/resolver.go [hunk 5]
```diff
@@ -345,6 +345,43 @@ func (r *Resolver) ResolveLinks() validation.ErrorList {
...
+				providerJob := providerInstanceGroup.LookupJob(consumer.JobName)
+				if providerJob == nil {
+					// This should not happen: we resolved a link, but can no
+					// longer find the job that provides it
+					errors = append(errors, validation.InternalError(
+						fmt.Sprintf(`instance_group[%s].job[%s].consumes[%s]`, consumerInstanceGroup.Name, consumerJob.Name, linkName),
+						fmt.Errorf(`Could not find resolved job %s in instance group %s`, consumer.JobName, consumer.RoleName)))
+					continue
+				}
+				if providerJob.ResolvedConsumedBy == nil {
```
Comment (f0rmiga) [inline model/resolver/resolver.go]: I think it should be initialized when the object is created.

File: model/resolver/validation.go [hunk 6]
```diff
@@ -54,6 +54,10 @@ func validateInstanceGroup(roleManifest *model.RoleManifest, g *model.InstanceGr
 			info.Name = name
 			jobReference.ResolvedConsumers[name] = info
 		}
+
+		if jobReference.ResolvedConsumedBy == nil {
```
Comment (f0rmiga) [inline model/resolver/validation.go:58]: Why a validation function is mutating a parameter passed to it? I see the previous code is doing it already, but it is way out of its concern doing so. Could you add a TODO for moving it out, and create a card for it on our board?

**Review by f0rmiga — CHANGES_REQUESTED** [review]

File: model/job_reference.go (hunk 1 above)
Comment (mook-as) [reply model/job_reference.go:14]: I feel like that should be done separately (this was just a comment change; refactoring would be a bigger change that should get its own PR).  I might still be convinced though!

Comment (f0rmiga) [reply model/job_reference.go:14]: It is ok to do in a separate PR, so we don't mix things up! Could we do right after?

File: model/resolver/validation.go (hunk 6 above)
Comment (f0rmiga) [reply model/resolver/validation.go:58]: Added a card to tackle this.

File: model/job_reference.go (hunk 1 above)
Comment (f0rmiga) [reply model/job_reference.go:14]: Card created to tackle this.

**Review by f0rmiga — APPROVED** [review]

**Pull request merged**